<a href="https://colab.research.google.com/github/connerdandrews/cse380-notebooks/blob/master/10_3_About_Patterns_and_Probabilities.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# About Patterns and Probabilities
## Class Directed Learning
### Due: Tuesday, 9 March 2021, 11:59 pm

## TODO Explore and Wonder

Regarding spanning trees of ladder graphs:

What is the probability ($\lim_{n \rightarrow \infty} P(n)$) that a randomly-selected spanning tree of an $n$-rung ladder graph contains the bottom rung?

| n | P(n) |
|---|------|
| 1 |      |
| 2 |      |
| 3 |      |
| 4 |      |
| 5 |      |


### Recurrence Relations

Let $f(n) =$ NSTIBR$(n)$:

$f(n) = 4f(n-1) - f(n-2)$ for $n > 1$;

$f(0) = 1$,

$f(1) = 1$.

Let $g(n) =$ NST$(n)$:

$g(n) = 4g(n-1) - g(n-2)$ for $n > 1$;

$g(0) = 0$,

$g(1) = 1$.

#### TODO Check Recurrences

Check the calculations in this table (maybe write recursive functions) to verify they are accurate.

Do they match what you found yesterday in your DPC?

In [18]:
def findBottomRungTreesInLadder(ladderLength):
  if ladderLength < 0:
    return
  if ladderLength == 0 or ladderLength == 1:
    return 1
  numberofTrees =  (4 * findBottomRungTreesInLadder(ladderLength - 1)) - findBottomRungTreesInLadder(ladderLength - 2)
  return numberofTrees

print(findBottomRungTreesInLadder(4))
print(findBottomRungTreesInLadder(5))
print(findBottomRungTreesInLadder(6))

41
153
571


| n | f(n) | f(n-1) | 4f(n-1)  | f(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    1 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      1 |        4 |      N/A |  N/A |
| 2 |    3 |      1 |        4 |        1 |    3 |
| 3 |   11 |      3 |       12 |        1 |   11 |
| 4 |   41 |     11 |       44 |        3 |   41 |
| 5 |  153 |     41 |      164 |       11 |  153 |

In [20]:
def findTreesInLadder(ladderLength):
  if ladderLength < 0:
    return
  if ladderLength == 0:
    return 0
  if ladderLength == 1:
    return 1
  numberofTrees =  (4 * findTreesInLadder(ladderLength - 1)) - findTreesInLadder(ladderLength - 2)
  return numberofTrees
  
print(findTreesInLadder(3))
print(findTreesInLadder(4))
print(findTreesInLadder(5))
print(findTreesInLadder(6))

15
56
209
780


| n | g(n) | g(n-1) | 4g(n-1)  | g(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    0 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      0 |        0 |      N/A |  N/A |
| 2 |    4 |      1 |        4 |        0 |    4 |
| 3 |   15 |      4 |       16 |        1 |   15 |
| 4 |   56 |     15 |       60 |        4 |   56 |
| 5 |  209 |     56 |      224 |       15 |  209 |

### TODO Find closed-form formulas

Can you find closed-form formulas for $f(n)$ and $g(n)$?

A closed-form formula expressing these functions in terms of operations on $n$, **without** referring to previous calculated values of the functions.

In [32]:
import math
from math import floor
from math import sqrt
def closedFormTrees(ladderLength):
  numberofTrees = ((1 / (2 * math.sqrt(3))) * ((2 + math.sqrt(3)) ** ladderLength)) -  ((1 / (2 * math.sqrt(3))) * ((2 - math.sqrt(3)) ** ladderLength))
  numberofTrees = floor(numberofTrees)
  return numberofTrees

print(closedFormTrees(3)) 
print(closedFormTrees(4)) 
print(closedFormTrees(5)) 
print(closedFormTrees(6)) 

15
56
209
780


#### Hint:

http://www.ist.tugraz.at/aichholzer/teaching/eca/spanning_trees_in_ladders.pdf

#### Possibly Illuminating Calculations

Remember continued fractions?

In [ ]:
from fractions import Fraction as frac

def contfrac2frac(seq):
    """Convert the simple continued fraction in `seq`
       into a fraction with numerator num and denominator den.
    """
    num, den = 1, 0
    for u in reversed(seq):
        num, den = den + num * u, num
    return frac(num, den)

def frac2contfrac(f):
    """Build the simple continued fraction expansion of fraction f.
    """
    seq = []
    frac2contfrac_rec(f, seq)
    return seq

def frac2contfrac_rec(f, seq):
    n = f.numerator
    d = f.denominator
    if d != 0:
        seq.append(n // d)
        if n % d != 0:
            frac2contfrac_rec(frac(d, n % d), seq)

def eval_frac(f):
    """Evaluate the fraction f as a float.
    """
    return f.numerator / f.denominator

In [ ]:
from math import sqrt

value = sqrt(3) - 1
value_as_cf = frac2contfrac(frac.from_float(value))[:21]
cf_to_value = contfrac2frac(value_as_cf)
print(value, value_as_cf, cf_to_value)
print(eval_frac(cf_to_value))

In [ ]:
from pprint import pprint

pprint(list(map(lambda n: contfrac2frac(value_as_cf[:n]), range(20, 5, -1))))